### DISCUSSION TITLE
Static cache

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#8105] [Ideas]
> Created by: Florian (fw@demodern.de) on 2015/03/13 06:38:29 +0000
> Votes at time of UserVoice import: 80

-----------------------------------

Besides the general fragment caching (and placing a varnish) infront of craft - I believe it would make sense to have the option of statically caching full pages as well. (i guess the majority of all pages with no user login's etc. will be fairly static. Most big cms'es render out there entire frontends as static html's to make sites super performance proof (i guess wordpress wont even work without W3T Cache or really static plugin.

When editing files within the CMS..resaving those page staticly also shouldnt be a biggy? Core or Plugin would be great for this.

**Comment by angrybrad** [comment]
> Posted by Johannes Lamers (johanneslamers@gmail.com) on 2015/03/10 17:03:25 +0000
-----------------------------------

This plugin is stellar! Will be part of my standard toolkit.

**Comment by angrybrad** [comment]
> Posted by Anonymous (chrstblt@gmail.com) on 2015/11/24 15:54:53 +0000
-----------------------------------

Hi Florian, some time has passed but there&#39;s a plugin available called HTMLCache (https://github.com/craftapi/htmlcache) that generates static HTML.

**Comment by MattiJarvinen** [comment]
I for one would love to see a Craft plugin writing static files straight into public folder thus removing the whole PHP part out of the picture.

What it would need in practice:

 - UTF-8 filesystem ( who doesn't really have this these days )
 - all slugs are handled as directories and index.html is used to store the actual page contents
 - Timed content expiry must be done with a cron job ( or similar )

The best thing would be also to include optional background regeneration/warming. When content entry is updated, instead of removing cached content files, operation would be as below.

```
for every url related to entry from DB cache
  trigger rerender
  store as static
```

This way craft cms would actually work mostly as a static site generator ( at least on parts where the content is cacheable ) resulting major speed up in performance.

**Comment by RitterKnightCreative** [comment]
Yeh would be nice if Craft did a very *minimal* version of static caching out of the box, outputting to a configurable folder that nginx/Apache would check before hitting the backend, eg like the way  [Statamic](https://statamic.dev/static-caching) and other CMS do it. 

I don't think we need something as full featured as Blitz with cache warming and all that or even going full featured as a static site generators like Craft's integration with gatsby but having a basic static cache built-in would solve a lot of (simpler) use cases where content doesn't change all that much. 

When we go to setup a section,  maybe there's a light switch  that allows caching for that section or turns it on/off by default. Then on the individual entries themselves,  there's a lightswitch to turn caching on or off for that page so dynamic content and forms wouldn't be affected. When the entry is saved, the cache is updated.

That's it, this would act more like a cold cache than anything.  The entries don't get busted until the entry is saved by the author. Very naive.

Let plugin authors extend the functionality in code events like ``onBeforeUpdateCache`` where they could define more advanced rules. I don't think Craft needs to go granular, like if the entry is in the news section, update that section index plus the home page and then trigger an upload to your netlifly CDN  or what have you.  That's exactly plugin territory but have the basic plumbing there. And have a console command to retrigger updates. That would allow developers to setup cron and make their own caching updates. 

I'm a fan of nginx fastcgi_cache but putting the cache generation in the hands of the CMS is even better IMO.
